apis/database/v1beta1/rdsinstance_types.go

```diff
@@ -0,0 +1,1160 @@
+	// +optional
+	Domain *string `json:"domain,omitempty"`
+
+	// DomainIAMRoleName specifes the name of the IAM role to be used when making API calls to the
+	// Directory Service.
+	// +optional
+	DomainIAMRoleName *string `json:"domainIAMRoleName,omitempty"`
+
+	// OptionGroupName indicates that the DB instance should be associated with the specified option
+	// group.
+	// Permanent options, such as the TDE option for Oracle Advanced Security TDE,
+	// can't be removed from an option group, and that option group can't be removed
+	// from a DB instance once it is associated with a DB instance
+	// +optional
+	OptionGroupName *string `json:"optionGroupName,omitempty"`
+
+	// A value that specifies that the DB instance class of the DB instance uses
+	// its default processor features.
+	UseDefaultProcessorFeatures *bool `json:"useDefaultProcessorFeatures,omitempty"`
+
+	// Determines whether a final DB snapshot is created before the DB instance
+	// is deleted. If true is specified, no DBSnapshot is created. If false is specified,
+	// a DB snapshot is created before the DB instance is deleted.
+	// Note that when a DB instance is in a failure state and has a status of 'failed',
+	// 'incompatible-restore', or 'incompatible-network', it can only be deleted
+	// when the SkipFinalSnapshotBeforeDeletion parameter is set to "true".
+	// Specify true when deleting a Read Replica.
+	// The FinalDBSnapshotIdentifier parameter must be specified if SkipFinalSnapshotBeforeDeletion
+	// is false.
+	// Default: false
+	SkipFinalSnapshotBeforeDeletion *bool `json:"skipFinalSnapshotBeforeDeletion,omitempty"`
+
+	// The DBSnapshotIdentifier of the new DBSnapshot created when SkipFinalSnapshot
+	// is set to false.
+	// Specifying this parameter and also setting the SkipFinalShapshot parameter
+	// to true results in an error.
+	// Constraints:
+	//    * Must be 1 to 255 letters or numbers.
+	//    * First character must be a letter
+	//    * Cannot end with a hyphen or contain two consecutive hyphens
+	//    * Cannot be specified when deleting a Read Replica.
+	FinalDBSnapshotIdentifier *string `json:"finalDBSnapshotIdentifier,omitempty"`
+}
+
+// An RDSInstanceSpec defines the desired state of an RDSInstance.
+type RDSInstanceSpec struct {
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+	ForProvider                  RDSInstanceParameters `json:"forProvider,omitempty"`
+
+	// DBSubnetGroupNameRef references to a DBSubnetGroup to retrieve its name
+	// +immutable
+	// +optional
+	DBSubnetGroupNameRef *DBSubnetGroupNameReferencerForRDSInstance `json:"dbsubnetGroupNameRef,omitempty" resource:"attributereferencer"`
+
+	// VPCSecurityGroupIDRefs references to a list of SecurityGroups to retrieve a list of securityGroupIDs
+	// +immutable
+	// +optional
+	VPCSecurityGroupIDRefs []*VPCSecurityGroupIDReferencerForRDSInstance `json:"vpcSecurityGroupIDRefs,omitempty" resource:"attributereferencer"`
```

Comment:
I've thought about this during[ the one-pager](https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#owner-based-struct-tree) and IMO `forProvider` should include only directly mirrored fields from SDK objects and act as the last representation of what's going to be sent.

But I'm happy to discuss changing that since at that time references were not implemented yet.

Reply:
I feel quite strongly that references belong under `forProvider`.

I can see the user benefit in `forProvider` being literally exactly what is sent to the provider, and moreso see the benefit that provides us when writing translation code. However, from a user's perspective reference fields _are_ for the provider - they're just a different form of specifying the input for the provider. I feel that keeping them immediately next to the underlying fields that they set makes them almost self-documenting. If you see `fooRef` right next to `foo` it's more obvious that the two fields are related than if they're far away from each other in the `spec`. Similarly, it maintains high fidelity for references.

Take the following example:

```yaml
apiVersion: example.org/v1
kind: Example
metadata:
  name: example
spec:
   # Does this field set .spec.forProvider.foo` or  .spec.forProvider.settings.foo?
   # Even outside of this pathological case in which two forProvider fields at
   # different levels have the same name, imagine adding a barRef here for a
   # bar field that was nested two levels deep inside forProvider in a complex
   # managed resource with tens of fields. It would be difficult to correlate them
   # at a glance.
   fooRef:
     name: example
  forProvider:
    foo: example
    settings:
      foo: example
```

Compared to:

```yaml
apiVersion: example.org/v1
kind: Example
metadata:
  name: example
spec:
  forProvider:
    foo: example
    fooRef:
      name: example
    settings:
      foo: example
      fooRef:
        name: example
```

I feel the latter example is much easier to understand.

I think we can avoid like issues with conflicting fields if we establish some rules, for example that such fields are always immediately next to their equivalent field and always named `fooRef` or `fooRefs` for arrays. We should probably also consider a comment marker for documentation purposes like `//+crossplane:reference:populates:foo`.